### PR TITLE
Propagate typeof() during inlining

### LINKED ISSUE #40381: RyuJIT: IsValueType and IsAssignableFrom optimizations aren't inlining friendly.
https://github.com/dotnet/runtime/pull/1157 introduced a jit optimization to optimize `typeof(X).IsValueType` to true/false. However, it seems this optimization doesn't work with inlining (same for `IsAssignableFrom`), e.g.: 

```csharp
bool Test() => Foo(typeof(int));

bool Foo(Type t1) => t1.IsValueType;
```
Codegen for `Test` (with `Foo` inlined):
```asm
G_M31255_IG01:
       4883EC28             sub      rsp, 40
G_M31255_IG02:
       48B9A09BC9ECF87F0000 mov      rcx, 0x7FF8ECC99BA0
       E82D238D5F           call     CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE
       488BC8               mov      rcx, rax
       48B8D8A8BEECF87F0000 mov      rax, 0x7FF8ECBEA8D8
       488B00               mov      rax, qword ptr [rax]
       3909                 cmp      dword ptr [rcx], ecx
G_M31255_IG03:
       4883C428             add      rsp, 40
       48FFE0               rex.jmp  rax
```
Expected codegen:
```asm
       mov      eax, 1   ;; return true
       ret  
```

the problem here is the fact that when we inline Foo we save `typeof(int)` into a temp variable (GT_LCL_VAR) and then importer is not able to optimize `get_IsValueType` because it expects input to be `typeof(X)` (`CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE`) rather than an unknown local, here is the IR after inlining:
```bash
# var tmp = typeof(x);
*  ASG       ref   
+--*  LCL_VAR   ref    V02 tmp1         
\--*  CALL help ref    HELPER.CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE
   \--*  CNS_INT(h) long   0x7ff8ecc99ba0 class


# return tmp.IsValueType;
*  CALL void   System.RuntimeType.IsValueTypeImpl
\--*  LCL_VAR   ref    V02 tmp1  
```
if we could, instead of saving `typeof(x)` into a temp propogate it by value during inlining I believe we'd get:
```bash
*  CALL void   System.RuntimeType.IsValueTypeImpl
\--*  CALL help ref    HELPER.CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE
   \--*  CNS_INT(h) long   0x7ff8ecc99ba0 class
```

Other options: early CSE ? or when I import `get_IsValueType` and see `LCL_VAR` arg with `lvSingleDef` flag - can I quickly get its assignment node?

PS: the same issue prevents the new API [Type.IsAssignableTo](https://github.com/dotnet/runtime/pull/40326) to be `typeof(X)` friendly since it uses `IsAssignableFrom` under the hood.

/cc @dotnet/jit-contrib 

category:cq
theme:inlining
skill-level:expert
cost:large

### PR DESCRIPTION
**Pull request by EgorBo**
Closes https://github.com/dotnet/runtime/issues/40381

Example:
```csharp
static void Main(string[] args)
{
    Console.WriteLine(Foo(typeof(float)));
}

static bool Foo(Type t)
{
    if (t == typeof(int))
        Console.WriteLine("1");
    if (t == typeof(uint))
        Console.WriteLine("2");
    if (t == typeof(long))
        Console.WriteLine("3");
    if (t == typeof(ulong))
        Console.WriteLine("4");
    return t.IsValueType;
}
```
Codegen for `Main()`:
```asm
       4883EC28             sub      rsp, 40
						;; size=4 bbWeight=1    PerfScore 0.25
G_M60419_IG02:              ;; offset=0004H
       B901000000           mov      ecx, 1
       FF15B1F43800         call     [System.Console:WriteLine(bool)]
       90                   nop      
						;; size=12 bbWeight=1    PerfScore 3.50
G_M60419_IG03:              ;; offset=0010H
       4883C428             add      rsp, 40
       C3                   ret  
```
It now:
1) Inliner now understands that `typeof()` at that call-site is a constant arg -> recognizes 4 foldabale branches in this inlinee:
```
Inline candidate has 4 foldable branches.  Multiplier increased to 8.
```
2) Propagates `typeof()` by value during inlining - it helps to eliminate those branches and fix https://github.com/dotnet/runtime/issues/40381

Regressions are due to inlining, RA and some cases where CSE/Hoisting doesn't "undo" propagation, e.g.: https://www.diffchecker.com/ukNlUweZ

Also a lot of regressions coming from the fact that we never do CSE across blocks, e.g.:
```csharp

static void Bar(int a) => Foo(a, typeof(int));

static void Foo(int a, Type t)
{
    if (a == 42)
        Console.WriteLine(t);
    else
        Console.WriteLine(t);
}
```
codegen diff for `Bar`: https://www.diffchecker.com/mWkBicM0
but it's not really a terrible perf regression I'd guess...

### CODE REVIEW

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -20206,6 +20206,17 @@ void Compiler::impInlineRecordArgInfo(InlineInfo*   pInlineInfo,
             return;
         }
     }
+    else
+    {
+        if (curArgVal->IsCall() &&
+            (curArgVal->AsCall()->gtCallMethHnd == eeFindHelper(CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE)) &&
+            (gtGetHelperArgClassHandle(curArgVal->AsCall()->gtArgs.GetArgByIndex(0)->GetEarlyNode()) !=
+             NO_CLASS_HANDLE))
```
Comment (EgorBo) [inline src/coreclr/jit/importer.cpp:20214]: ^ will be replaced with `gtIsTypeof` from https://github.com/dotnet/runtime/pull/71685/files

**Comment by EgorBo** [comment]
PTAL @dotnet/jit-contrib @AndyAyersMS
overall, [good diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1870182&view=ms.vss-build-web.run-extensions-tab) size regressions mostly due to the fact we propagate `typeof()` into different basic blocks and then don't CSE it back - but it might be a perf improvement actually, because we don't call the helper when it's not needed.

File: src/coreclr/jit/gentree.cpp [hunk 2]
```diff
@@ -12913,17 +12913,19 @@ GenTree* Compiler::gtFoldTypeCompare(GenTree* tree)
...
-        return tree;
-    }
-
     GenTree* const         op2     = tree->AsOp()->gtOp2;
+    const TypeProducerKind op1Kind = gtGetTypeProducerKind(op1);
     const TypeProducerKind op2Kind = gtGetTypeProducerKind(op2);
-    if (op2Kind == TPK_Unknown)
+
+    // Fold "typeof(handle) cmp null"
+    if ((op2->IsIntegralConst(0) && (op1Kind == TPK_Handle)) || (op1->IsIntegralConst(0) && (op2Kind == TPK_Handle)))
```
Comment (AndyAyersMS) [inline src/coreclr/jit/gentree.cpp:12921]:
```suggestion
    if (((op2Kind == TPK_Null) && (op1Kind == TPK_Handle)) || ((op1Kind == TPK_ Null) && (op2Kind == TPK_Handle)))
```

Comment (AndyAyersMS) [reply src/coreclr/jit/gentree.cpp:12921]: Doesn't `gtGetTypeProducerKind` already look for nulls?

Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp]: Ah, indeed, fixed

**Comment by EgorBo** [comment]
@AndyAyersMS anything else?

**Review by AndyAyersMS — APPROVED** [review]

**Pull request merged**